File: nr_robot/Automation/src/detection.py
#!/usr/bin/env python3

import rospy
import tf
import numpy as np
import franka_gripper.msg
import sys
import actionlib
import time
import numpy as np

from geometry_msgs.msg import PoseStamped, Quaternion
from franka_msgs.msg import FrankaState
from tf.transformations import *
import tf.transformations as tr

from yolov5_ros.srv import petri, petriResponse

from visualization_msgs.msg import Marker
from geometry_msgs.msg import Quaternion, Pose, Point, Vector3
from std_msgs.msg import Header, ColorRGBA


initial_pose = PoseStamped()
initial_pose_found = False
pose_to_pub = PoseStamped()
pose_to_pub2 = PoseStamped()

def show_text_in_rviz(pos_temp):
    marker_publisher.publish(pos_temp)

def pos_to_mat(quat,trasl):
	qx = quat[0]
	qy = quat[1]
	qz = quat[2]
	qw = quat[3]
	tx = trasl[0]
	ty = trasl[1]
	tz = trasl[2]
	q = [qx, qy, qz, qw]
	t = [tx, ty, tz]
	T_r = tr.quaternion_matrix(q)
	T_t  =tr.translation_matrix(t)
	T = T_r + T_t - [[1, 0, 0 , 0],[0, 1, 0 , 0],[0, 0, 1 , 0],[0, 0, 0 , 1]]
	return T


def pos_to_mat_1(x_camera, y_camera, z_camera):
	qx = 0#pos.pose.orientation.x
	qy = 0#pos.pose.orientation.y
	qz = 0#pos.pose.orientation.z
	qw = 1#pos.pose.orientation.w
	tx = x_camera#pos.pose.position.x
	ty = y_camera#pos.pose.position.y
	tz = z_camera#pos.pose.position.z
	q = [qx, qy, qz, qw]
	t = [tx, ty, tz]
	T_r = tr.quaternion_matrix(q)
	T_t  =tr.translation_matrix(t)
	T = T_r + T_t - [[1, 0, 0 , 0],[0, 1, 0 , 0],[0, 0, 1 , 0],[0, 0, 0 , 1]]
	return T


# da matrix a pos

def mat_to_pos(mat,frame_name):
	pos = PoseStamped()
	pos.pose.position.x = mat[0][3]
	pos.pose.position.y = mat[1][3]
	pos.pose.position.z = mat[2][3]
	q_temp = tr.quaternion_from_matrix(mat)
	q = tr.unit_vector(q_temp)
	pos.pose.orientation.x = q[0]
	pos.pose.orientation.y = q[1]
	pos.pose.orientation.z = q[2]
	pos.pose.orientation.w = q[3]
	pos.header.stamp = rospy.Time.now()
	pos.header.frame_id = frame_name
	return pos


def franka_state_callback(msg):
    initial_quaternion = \
        tf.transformations.quaternion_from_matrix(
            np.transpose(np.reshape(msg.O_T_EE,
                                    (4, 4))))
    initial_quaternion = initial_quaternion / np.linalg.norm(initial_quaternion)


    initial_pose.pose.orientation.x = initial_quaternion[0]
    initial_pose.pose.orientation.y = initial_quaternion[1]
    initial_pose.pose.orientation.z = initial_quaternion[2]
    initial_pose.pose.orientation.w = initial_quaternion[3]
    initial_pose.pose.position.x = msg.O_T_EE[12]
    initial_pose.pose.position.y = msg.O_T_EE[13]
    initial_pose.pose.position.z = msg.O_T_EE[14]
    global initial_pose_found
    initial_pose_found = True
    print(initial_pose)

def movement(speed, pos, orient):
    start_pose = initial_pose
    start_time = rospy.get_rostime()

    num_steps = int(speed* 10)
    step_size = 1.0 / num_steps

    if len(orient) == 3:
        rotation_matrix = tf.transformations.euler_matrix( orient[0], orient[1], orient[2])
        goal_quat = quaternion_from_matrix(rotation_matrix)
        orient  =  goal_quat
    # orient = Quaternion(orient[0], orient[1], orient[2], orient[3])
    start_orient = [start_pose.pose.orientation.x, start_pose.pose.orientation.y, start_pose.pose.orientation.z, start_pose.pose.orientation.w]
    print(orient)
    print(start_orient)
    orient_new = quaternion_multiply(orient, start_orient)

    old_x = 0
    old_y = 0
    old_z = 0

    for i in range(num_steps):
        t = float(i) / (num_steps-1)
        pose_to_pub.pose.position.x =  (1 - t) * start_pose.pose.position.x + t * pos[0]
        pose_to_pub.pose.position.y = (1 - t) * start_pose.pose.position.y + t *  pos[1]
        pose_to_pub.pose.position.z =   (1 - t) * start_pose.pose.position.z + t * pos[2]


        pose_to_pub.pose.orientation.x = orient_new[0]
        pose_to_pub.pose.orientation.y = orient_new[1]
        pose_to_pub.pose.orientation.z = orient_new[2]
        pose_to_pub.pose.orientation.w = orient_new[3]


        pose_to_pub.header.frame_id = "panda_link0"
        pose_to_pub.header.stamp = rospy.Time.now()
        pose_pub.publish(pose_to_pub)

        print("----pose published-----")
        print(pose_to_pub)
        print("----end pose published-----")
        if pose_to_pub.pose.position.x == old_x and pose_to_pub.pose.position.y == old_y and pose_to_pub.pose.position.z == old_z:
            print("stopped")
            break


        pose_to_pub.pose.position.x = old_x
        pose_to_pub.pose.position.y = old_y
        pose_to_pub.pose.position.z = old_z



    # rospy.sleep(2)
    print("----state callback-----")
    print(initial_pose)
    print("----end state callback-----")





    # print(pose_to_pub.pose)
    # show_text_in_rviz(pose_to_pub)

def detect(x_camera, y_camera, z_camera):
    listener = tf.TransformListener()
    # publishing to equilibrium pose only working with panda_link0
    # so we change end effector tansform to panda_link0
    listener.waitForTransform('panda_link0', 'camera_color_optical_frame', rospy.Time(0), rospy.Duration(4.0))
    trans, rot = listener.lookupTransform('panda_link0', 'camera_color_optical_frame', rospy.Time(0))


    T_target_pose_detection_1 = pos_to_mat_1(x_camera, y_camera, z_camera)
    T_1 = pos_to_mat(rot,trans)
    T_result_temp = np.dot(T_1,T_target_pose_detection_1)
    pos_temp = mat_to_pos(T_result_temp,"panda_link0")
    print (pos_temp.pose.position.x,pos_temp.pose.position.y, pos_temp.pose.position.z)
    return [pos_temp.pose.position.x,pos_temp.pose.position.y,pos_temp.pose.position.z]

def box(petri_Detect, pos_home, quat_home, pos_detect_home, quat_detect_home):
    x_camera = -0.06027261510178125
    y_camera = 0.025831120775062503
    z_camera = 0.0
    num = 1
    # zone
    for i in range(num):
        # pick
        resp = petri_Detect(1)
        # print service call response to screen
        print(resp.result)
        x_camera = resp.result[0]
        y_camera = resp.result[1]
        z_camera = resp.result[2]
        arr = detect(x_camera, y_camera, z_camera)
        pos_t = (arr[0], arr[1],  arr[2])#0.40)
        quat_t= (0.3622237207012903, 0.9320895572595912, 0.0016233448866874078,-0.0006309994548575625)
        movement(2000, pos_t, quat_t)
        pos_tilt = (arr[0], arr[1],  arr[2])#0.19)
        quat_tilt= (0.3698165170753044, 0.9288548265726251, 0.00821662343612614,-0.019923402077426496)
        # quat_tilt= (0.39472459160942525,  0.864290503089696, 0.11103097811698952, 0.2913186313061995)
        movement(5000, pos_tilt, quat_tilt)
        rospy.sleep(1)

        # pick rise 1st part
        pos_t = (arr[0], arr[1],  0.80)
        movement(4000, pos_t, quat_t)

        # # # # pick rise 2nd part
        # # # pos_t = (arr[0], arr[1],  0.950)
        # # # quat_t = (0.3969233021430672, 0.7554622198723355, 0.18641445824799882, 0.4868042484587736)
        # # # movement(3500, pos_t, quat_t)
        #
        # # place mid point
        # pos_place= (0.1717892411578945, 0.3253767856244797, 0.9515217903369477)
        # quat_place= (3.14159265359, 0, 2.00)
        # movement(3000, pos_place, quat_place)
        #
        # # place directly above
        # pos_place= (-0.2140738234, 0.4663263380584926, 0.920098406)
        # quat_place= (3.14159265359, 0, 3.00)#1.57)
        # movement(3000, pos_place, quat_place)
        #
        # # place directly settled
        #bd_box_z  pos_place= (-0.2140738234, 0.4759754556, 0.5)
        # quat_place= (3.14159265359, 0.0, 0.8)
        # movement(7000, pos_place, quat_place)
        # rospy.sleep(1)
        #
        # home
        # movement(4000, pos_home, quat_home)
        movement(4000, pos_detect_home, quat_detect_home)
        # if i != num-1:
        #     movement(3000, pos_detect_home, quat_detect_home)
        #     rospy.sleep(1)




if __name__ == "__main__":
    try:
        rospy.init_node('detection')
        rate = rospy.Rate(500.0)

        rospy.wait_for_service('/petri_service')

        state_sub = rospy.Subscriber("franka_state_controller/franka_states",
                                    FrankaState, franka_state_callback)
        pose_pub = rospy.Publisher(
            "cartesian_impedance_example_controller/equilibrium_pose",  PoseStamped, queue_size=100)
        marker_publisher = rospy.Publisher('marker', PoseStamped, queue_size=10)

        # initialise service proxy
        petri_Detect = rospy.ServiceProxy('/petri_service', petri)

        connections = pose_pub.get_num_connections()
        rospy.loginfo('Connections: %d', connections)

        rospy.sleep(2)

        # cordinates and orientations
        #home orientation:
        pos_home = (0.06885265922638234,0.007600406292727626,0.675742975736092)
        quat_home = (0.39051698780590594, 0.9202975564500481,0.01006373672315757,0.021156819946287313)

        #detect_home
        pos_detect_home = ( 0.2728995830087986, -0.03516536460051363, 0.8374402976709023)
        # quat_detect_home= (0.37938161845937964, 0.9144142500555361, -0.024811024453374644, -0.05103536022431737)
        quat_detect_home = (3.14159265359, 1.5, 4.2)

        # pos_detect_home = ( 0.12904655926727843,  -0.045988363016644655, 0.9104944218262105)
        # quat_detect_home= (0.36963400892159226,  0.9229728799315546, 0.04164122417637632, 0.09878142947182361)
        # pos_detect_home = ( 0.3824513267425953,-0.1578228798153445,0.7507923156710219)
        # quat_detect_home= (0.37685686373845484, 0.9261915695034868, -0.007020554390991775,0.009939449176225468)

        while not initial_pose_found:
            rospy.sleep(1)
        # show_text_in_rviz(pos_temp)
        # rospy.sleep(2)


        # home
        movement(2000, pos_home, quat_home)

        # detect home
        movement(2000, pos_detect_home, quat_detect_home)
        rospy.sleep(1)

        # box process
        # box(petri_Detect, pos_home, quat_home, pos_detect_home, quat_detect_home)

        rospy.spin()

    except rospy.ROSInterruptException:
	    pass

File: nr_robot/Automation/src/detection.cpp
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>

#include <moveit_msgs/DisplayRobotState.h>
#include <moveit_msgs/DisplayTrajectory.h>


#include <moveit/robot_model_loader/robot_model_loader.h>
#include <moveit/robot_model/robot_model.h>
#include <moveit/robot_state/robot_state.h>

#include <moveit_visual_tools/moveit_visual_tools.h>

#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf/transform_listener.h>
#include <geometry_msgs/PointStamped.h>

#include <iostream>
#include <vector>

#include "yolov5_ros/petri.h"
#include <cstdlib>

// void interpolate (){
//   ros::init(argc, argv, "pose_publisher");
//   ros::NodeHandle nh;
//   ros::Publisher pose_pub = nh.advertise<geometry_msgs::PoseStamped>("pose_topic", 1);

//   geometry_msgs::PoseStamped start_pose;
//   // Initialize start_pose and other variables as needed

//   ros::Time start_time = ros::Time::now();

//   double speed = 1.0;  // Set your desired speed
//   int num_steps = static_cast<int>(speed * 10);
//   double step_size = 1.0 / num_steps;

//   tf::Quaternion orient;
//   if (orient.size() == 3)
//   {
//       tf::Matrix3x3 rotation_matrix;
//       rotation_matrix.setEulerYPR(orient[0], orient[1], orient[2]);
//       tf::Quaternion goal_quat;
//       rotation_matrix.getRotation(goal_quat);
//       orient = goal_quat;
//   }

//   tf::Quaternion start_orient(start_pose.pose.orientation.x, start_pose.pose.orientation.y,
//                               start_pose.pose.orientation.z, start_pose.pose.orientation.w);

//   tf::Quaternion orient_new = orient * start_orient;

//   double old_x = 0.0;
//   double old_y = 0.0;
//   double old_z = 0.0;

//   ros::Rate rate(10);  // Publish at 10 Hz

//   for (int i = 0; i < num_steps; ++i)
//   {
//       double t = static_cast<double>(i) / (num_steps - 1);
//       geometry_msgs::PoseStamped pose_to_pub;

//       pose_to_pub.pose.position.x = (1.0 - t) * start_pose.pose.position.x + t * pos[0];
//       pose_to_pub.pose.position.y = (1.0 - t) * start_pose.pose.position.y + t * pos[1];
//       pose_to_pub.pose.position.z = (1.0 - t) * start_pose.pose.position.z + t * pos[2];

//       pose_to_pub.pose.orientation.x = orient_new.x();
//       pose_to_pub.pose.orientation.y = orient_new.y();
//       pose_to_pub.pose.orientation.z = orient_new.z();
//       pose_to_pub.pose.orientation.w = orient_new.w();

//       pose_to_pub.header.frame_id = "panda_link0";
//       pose_to_pub.header.stamp = ros::Time::now();
//       pose_pub.publish(pose_to_pub);

//       ROS_INFO("----pose published-----");
//       ROS_INFO_STREAM(pose_to_pub);
//       ROS_INFO("----end pose published-----");

//       if (pose_to_pub.pose.position.x == old_x && pose_to_pub.pose.position.y == old_y &&
//           pose_to_pub.pose.position.z == old_z)
//       {
//           ROS_INFO("stopped");
//           break;
//       }

//       old_x = pose_to_pub.pose.position.x;
//       old_y = pose_to_pub.pose.position.y;
//       old_z = pose_to_pub.pose.position.z;

//       rate.sleep();
//   }

// }

void movement (moveit::planning_interface::MoveGroupInterface& move_group,float& x,float& y,float& z, float& x_D,float& y_D,float& z_D) {
  geometry_msgs::Pose target_pose1;
  tf2::Quaternion orientation;
  orientation.setRPY(x_D, y_D, z_D);
  target_pose1.orientation = tf2::toMsg(orientation);
  // target_pose1.orientation.w = 0;
  target_pose1.position.x = x;
  target_pose1.position.y = y;
  target_pose1.position.z = z;
  move_group.setPoseTarget(target_pose1);
  moveit::planning_interface::MoveGroupInterface::Plan my_plan;

  bool success = (move_group.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
  ROS_INFO_NAMED("tutorial", "Visualizing plan 1 (pose goal) %s", success ? "" : "FAILED");
  std::cout << target_pose1;
  move_group.execute(my_plan);
  // move_group.move();
}

std::vector<float> detect (std::vector<double> camera_frame) {
  tf::TransformListener listener;
  std::cout << "camera optical frame: ";
  for (const auto &value : camera_frame) {
        std::cout << value << " ";
    }
  std::cout << std::endl;
  // Wait for the transform from 'panda_link0' to 'camera_color_optical_frame'
  ros::Duration timeout(2.0);
  if (!listener.waitForTransform("panda_link0", "camera_color_optical_frame", ros::Time(0), timeout))
  {
      ROS_ERROR("Transform from 'panda_link0' to 'camera_color_optical_frame' not available.");
      std::vector<float> error_result(3, 0.0); // Return a vector of zeros
      return error_result;
  }
  float object_x = camera_frame[0];  // Replace with your object's coordinates
  float object_y = camera_frame[1];
  float object_z = camera_frame[2];

  // Create a PointStamped message to represent the object's position
  geometry_msgs::PointStamped camera_point;
  camera_point.header.frame_id = "camera_color_optical_frame";
  camera_point.header.stamp = ros::Time(0);
  camera_point.point.x = object_x;
  camera_point.point.y = object_y;
  camera_point.point.z = object_z;

  geometry_msgs::PointStamped panda_point;

  try
  {
      listener.transformPoint("panda_link0", camera_point, panda_point);
  }
  catch (tf::TransformException& ex)
  {
      ROS_ERROR("Failed to transform point: %s", ex.what());
      std::vector<float> error_result(3, 0.0); // Return a vector of zeros
      return error_result;
  }
  std::vector<float> arr(3);
  // Extract the transformed point in panda_link0 coordinates
  arr[0] = panda_point.point.x;
  arr[1] = panda_point.point.y;
  arr[2] = panda_point.point.z;
  // You can now use the 'object_x_panda', 'object_y_panda', and 'object_z_panda' variables
  // as the object's coordinates in panda_link0 coordinates.

  return arr;
}


void addCollisionObjects(moveit::planning_interface::PlanningSceneInterface& planning_scene_interface)
{
  // BEGIN_SUB_TUTORIAL table1
  //
  // Creating Environment
  // ^^^^^^^^^^^^^^^^^^^^
  // Create vector to hold 3 collision objects.
  std::vector<moveit_msgs::CollisionObject> collision_objects;
  collision_objects.resize(3);

  // Add the first table where the cube will originally be kept.
  collision_objects[0].id = "table1";
  collision_objects[0].header.frame_id = "panda_link0";

  /* Define the primitive and its dimensions. */
  collision_objects[0].primitives.resize(1);
  collision_objects[0].primitives[0].type = collision_objects[0].primitives[0].BOX;
  collision_objects[0].primitives[0].dimensions.resize(3);
  collision_objects[0].primitives[0].dimensions[0] = 0.2;
  collision_objects[0].primitives[0].dimensions[1] = 0.6;
  collision_objects[0].primitives[0].dimensions[2] = 1.2;

  /* Define the pose of the table. */
  collision_objects[0].primitive_poses.resize(1);
  collision_objects[0].primitive_poses[0].position.x = -0.6;
  collision_objects[0].primitive_poses[0].position.y = 0;
  collision_objects[0].primitive_poses[0].position.z = 0.6;
  // END_SUB_TUTORIAL

  collision_objects[0].operation = collision_objects[0].ADD;

  // BEGIN_SUB_TUTORIAL table2
  // Add the second table where we will be placing the cube.
  collision_objects[1].id = "table2";
  collision_objects[1].header.frame_id = "panda_link0";

  /* Define the primitive and its dimensions. */
  collision_objects[1].primitives.resize(1);
  collision_objects[1].primitives[0].type = collision_objects[1].primitives[0].BOX;
  collision_objects[1].primitives[0].dimensions.resize(3);
  collision_objects[1].primitives[0].dimensions[0] = 0.4;
  collision_objects[1].primitives[0].dimensions[1] = 0.2;
  collision_objects[1].primitives[0].dimensions[2] = 1.2;

  /* Define the pose of the table. */
  collision_objects[1].primitive_poses.resize(1);
  collision_objects[1].primitive_poses[0].position.x = 0;
  collision_objects[1].primitive_poses[0].position.y = -0.4;
  collision_objects[1].primitive_poses[0].position.z = 0.6;
  // END_SUB_TUTORIAL

  collision_objects[1].operation = collision_objects[1].ADD;

  // BEGIN_SUB_TUTORIAL table2
  // Add the second table where we will be placing the cube.
  collision_objects[2].id = "table3";
  collision_objects[2].header.frame_id = "panda_link0";

  /* Define the primitive and its dimensions. */
  collision_objects[2].primitives.resize(1);
  collision_objects[2].primitives[0].type = collision_objects[1].primitives[0].BOX;
  collision_objects[2].primitives[0].dimensions.resize(3);
  collision_objects[2].primitives[0].dimensions[0] = 1.4;
  collision_objects[2].primitives[0].dimensions[1] = 0.4;
  collision_objects[2].primitives[0].dimensions[2] = 0.6;

  /* Define the pose of the table. */
  collision_objects[2].primitive_poses.resize(1);
  collision_objects[2].primitive_poses[0].position.x = 0.7;
  collision_objects[2].primitive_poses[0].position.y = 0.6;
  collision_objects[2].primitive_poses[0].position.z = 0.3;
  // END_SUB_TUTORIAL

  collision_objects[2].operation = collision_objects[2].ADD;


  planning_scene_interface.applyCollisionObjects(collision_objects);
}




int main(int argc, char** argv)
{
  ros::init(argc, argv, "detection");
  ros::NodeHandle node_handle;
  ros::AsyncSpinner spinner(1);
  spinner.start();
  ros::ServiceClient client = node_handle.serviceClient<yolov5_ros::petri>("/petri_service");
  static const std::string PLANNING_GROUP = "panda_arm";

  moveit::planning_interface::MoveGroupInterface move_group(PLANNING_GROUP);

  moveit::planning_interface::PlanningSceneInterface planning_scene_interface;

  move_group.setMaxVelocityScalingFactor(0.1);
  move_group.setMaxAccelerationScalingFactor(0.05);

  // addCollisionObjects(planning_scene_interface);

  robot_model_loader::RobotModelLoader robot_model_loader("robot_description");
  const moveit::core::RobotModelPtr& kinematic_model = robot_model_loader.getModel();
  
  moveit::core::RobotStatePtr kinematic_state(new moveit::core::RobotState(kinematic_model));
  kinematic_state->setToDefaultValues();
  const moveit::core::JointModelGroup* joint_model_group = kinematic_model->getJointModelGroup("panda_arm");
  
  kinematic_state->setToRandomPositions(joint_model_group);
  const Eigen::Isometry3d& end_effector_state = kinematic_state->getGlobalLinkTransform("panda_link0");
  
  /* Print end-effector pose. Remember that this is in the model frame */
  ROS_INFO_STREAM("Translation: \n" << end_effector_state.translation() << "\n");
  ROS_INFO_STREAM("Rotation: \n" << end_effector_state.rotation() << "\n");

  for (int i = 0; i < 3; i++){
    // home
    float x,y,z, x_D, y_D, z_D;
    // x = 0.25;
    // y = 0.0;
    // z = 0.8;
    x = 0.3;
    y = 0.0;
    z = 0.8;
    x_D = 0;
    y_D = 3.926;//3.142;
    z_D = 0;//5.49;
    movement(move_group, x, y, z, x_D, y_D, z_D);

    // // transversals operations
    // ros::Duration(2).sleep(); // sleep for camera detection
    // yolov5_ros::petri srv;
    // int detect_var = 1;
    // srv.request.detect = detect_var;
    // if (client.call(srv))
    // {
    //   // pick above operation
    //   ROS_INFO("Object in camera optical frame called");
    //   std::vector<double> transformedPoints(srv.response.result.begin(), srv.response.result.end());
    //   std::vector<float> transformedPointsPanda = detect(transformedPoints);
    //   ROS_INFO("Object in pandalink0 frame found");
    //   for (const auto &value : transformedPointsPanda) {
    //         std::cout << value << " ";
    //     }
    //   std::cout << std::endl;
    //   float x = transformedPointsPanda[0];
    //   float y = transformedPointsPanda[1];
    //   float z = 0.7;//transformedPointsPanda[2];
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // pick operation 1
    //   z = 0.61;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // pick operation 2
    //   z = 0.31;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // pick operation 3
    //   z = 0.28;//0.26;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);
    //   ros::Duration(2).sleep();

    //   // pick rise operation 1
    //   z = 0.40;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // pick rise operation 2
    //   z = 0.65;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // pick rise operation 3
    //   z = 0.80;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // place midpoint operation 1
    //   y = 0.37;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // place midpoint operation 2
    //   x = -0.19;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // place directly above operation
    //   x = -0.214;
    //   y = 0.4663;
    //   z = 0.80;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // place drop 1 operation
    //   x = -0.214;
    //   y = 0.4663;
    //   z = 0.62;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);

    //   // place drop 2 operation
    //   x = -0.214;
    //   y = 0.4663;
    //   z = 0.42;
    //   movement(move_group, x, y, z, x_D, y_D, z_D);
    // } else {
    //   ROS_ERROR("failed to call service");
    //   return 1;
    // }

  }




  ros::waitForShutdown();
  return 0;
}
